Clover/app/src/main/java/org/floens/chan/core/repository/ImportExportRepository.java

```diff
@@ -0,0 +1,476 @@
+import org.floens.chan.core.model.orm.ThreadHide;
+import org.floens.chan.core.settings.ChanSettings;
+import org.floens.chan.utils.Logger;
+
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import javax.inject.Inject;
+
+public class ImportExportRepository {
+    private static final String TAG = "ImportExportRepository";
+    public static final String EXPORT_FILE_NAME = "exported_clover_settings.json";
+
+    // Don't forget to change this when changing any of the Export models.
+    // Also, don't forget to handle the change in the onUpgrade or onDowngrade methods
+    public static final int CURRENT_EXPORT_SETTINGS_VERSION = 1;
+
+    private DatabaseManager databaseManager;
+    private DatabaseHelper databaseHelper;
+    private Gson gson;
+
+    @Inject
+    public ImportExportRepository(
+            DatabaseManager databaseManager,
+            DatabaseHelper databaseHelper,
+            Gson gson
+    ) {
+        this.databaseManager = databaseManager;
+        this.databaseHelper = databaseHelper;
+        this.gson = gson;
+    }
+
+    public void exportTo(File settingsFile, ImportExportCallbacks callbacks) {
+        databaseManager.runTask(() -> {
+            try {
+                ExportedAppSettings appSettings = readSettingsFromDatabase();
+                if (appSettings.isEmpty()) {
+                    callbacks.onNothingToImportExport(ImportExport.Export);
+                    return null;
+                }
+
+                String json = gson.toJson(appSettings);
+
+                if (settingsFile.exists()) {
+                    if (!settingsFile.isFile()) {
+                        throw new IOException(
+                                "Settings file is not a file (???) " + settingsFile.getAbsolutePath()
+                        );
+                    }
+
+                    if (!settingsFile.delete()) {
+                        Logger.w(TAG, "Could not delete export file before exporting " + settingsFile.getAbsolutePath());
+                    }
+                }
+
+                if (!settingsFile.createNewFile()) {
+                    throw new IOException(
+                            "Could not create a file for exporting " + settingsFile.getAbsolutePath()
+                    );
+                }
+
+                if (!settingsFile.exists() || !settingsFile.canWrite()) {
+                    throw new IOException(
+                            "Something wrong with export file (Can't write or it doesn't exist) "
+                                    + settingsFile.getAbsolutePath()
+                    );
+                }
+
+                try (RandomAccessFile raf = new RandomAccessFile(settingsFile, "rw")) {
```

Comment:
This doesn't work correctly with UTF-8 encoded strings. Some symbols tend to get fucked up and then you won't be able to import afterwards. Use a FileWriter instead, which writes out UTF-8 by default and doesn't cause issues later on.

Reply:
Yeah, apparently for UTF to work it should be writeUtf() not writeBytes().